includes/admin/class-wc-rest-payments-webhook-controller.php

```diff
@@ -229,6 +232,36 @@ private function process_webhook_expired_authorization( $event_body ) {
 		WC_Payments_Utils::mark_payment_expired( $order );
 	}
 
+	/**
+	 * Process webhook for a successul payment intent.
+	 *
+	 * @param array $event_body The event that triggered the webhook.
+	 *
+	 * @throws Rest_Request_Exception           Required parameters not found.
+	 * @throws Invalid_Payment_Method_Exception When unable to resolve charge ID to order.
+	 */
+	private function process_webhook_payment_intent_succeeded( $event_body ) {
+		$event_data   = $this->read_rest_property( $event_body, 'data' );
+		$event_object = $this->read_rest_property( $event_data, 'object' );
+
+		$intent_id = $this->read_rest_property( $event_object, 'id' );
+
+		// Look up the order related to this charge.
+		$order = $this->wcpay_db->order_from_intent_id( $intent_id );
+		if ( ! $order ) {
+			throw new Invalid_Payment_Method_Exception(
+				sprintf(
+					/* translators: %1: charge ID */
+					__( 'Could not find order via intent ID: %1$s', 'woocommerce-payments' ),
+					$intent_id
+				),
+				'order_not_found'
+			);
+		}
+
+		$order->payment_complete();
```

Comment:
It's not impossible that this webhook could be fired multiple times. It's unlikely, but if there's an easy way to guard against it we should do it.

What happens if `payment_complete` is called on an already complete order?

Reply:
Thanks for calling this out!

I can do something like:
```
if ( ! $order->has_status( array( 'processing', 'completed' ) ) ) {
			$order->payment_complete();
}
```

> What happens if payment_complete is called on an already complete order?

To verify, I tested checkout with a `completed` virtual order and a `processing` physical order separately. I manually fire the same `payment_intent.succeeded` event for each of them. I observed both of them skipping into the `else` condition [here] (https://github.com/woocommerce/woocommerce/blob/aab56d84e00f34bf8bfc8634f8c7aff9ae8c7ca6/includes/class-wc-order.php#L122-L124) as expected. None of these orders are re-processed again.

Out of curiosity I took a little deeper dive into WooCommerce core. It seems like `payment_complete` [already has the above logic](https://github.com/woocommerce/woocommerce/blob/aab56d84e00f34bf8bfc8634f8c7aff9ae8c7ca6/includes/class-wc-order.php#L111) so that core does not set `processing`/`completed` status again. At first glance, it looks like we don't need to implement the same logic again. However, I noticed core's `payment_complete()` also call a few other hooks. It is possible that other plugins might be relying on these hooks. Thus, our retry webhook here can trigger these functions again and we don't know what they would do.

Given the above, although it is nice to rely on core to [control statuses](https://github.com/woocommerce/woocommerce/blob/aab56d84e00f34bf8bfc8634f8c7aff9ae8c7ca6/includes/class-wc-order.php#L111), I do agree that it is safer to not call `$order-> payment_complete()` unless the order is not already `processed`/`completed`.

Reply:
Fixed in 9653c8f5137cb2143bc0369807307d5aad6ace75 and added a test case.